Integrate secondary analysis information to DCP-Diag

rexwangcc opened the pull request:
## For reviewers
### This PR did the following things:
- Refactor most of (not all of) the import statements to use absolute imports(Python officially prefers absolute than relative) than relative imports(which caused a lot of issues to me during the development process).
- Create a new directory ` dcp_diag/component_entities` and migrate all of the entities previously living under `dcp_diag/component_agents` to their new home.
- Implement the analysis agent and analysis entities. The former is the agent/client who talks to secondary analysis, the latter is an abstract model of Analysis Workflows.
- Apply Google style docstrings to all of the analysis-related code, which could then be converted to readthedocs.io documentation automatically with `Spinx`.
- Update requirements to support analysis agent.
- Implement analysis finder and the logic that queries for analysis entities in `dcpdig` command.
- Implement the logic that queries for analysis entities by `project UUID` in `analyze-submission` command.
- Modify the structure of the persistent state JSON file, which will look like the following:
```JSON
{
    "version": 1,
    "submission_id": "sss",
    "project_uuid": "eee",
    "bundle_map": {
        "aaa": {
            "type": "primary",
            "aws": {
                "dss_presence": true,
                "in_dss_project_search": true,
                "results_bundles": [
                    "aaa1.yyy"
                ]
            },
            "gcp": {
                "dss_presence": true,
                "in_dss_project_search": true,
                "results_bundles": [
                    "aaa1.yyy"
                ]
            },
            "fqid": "aaa.yyy",
            "analysis_workflows": {
                "ZZZ": {
                    "end": "2019-01-06T20:35:19.533Z",
                    "id": "ZZZ",
                    "labels": {
                        "bundle-uuid": "aaa",
                        "bundle-version": "yyy",
                        "caas-collection-name": "lira-prod",
                        "cromwell-workflow-id": "cromwell-ZZZ",
                        "project_shortname": "project name",
                        "project_uuid": "eee",
                        "workflow-name": "AdapterSmartSeq2SingleCell",
                        "workflow-version": "smartseq2_v2.1.0"
                    },
                    "name": "AdapterSmartSeq2SingleCell",
                    "start": "2019-01-06T20:18:18.102Z",
                    "status": "Succeeded",
                    "submission": "2019-01-06T20:16:30.804Z"
                },
                "ZZZ2": {
                    "end": "2019-01-06T20:35:19.533Z",
                    "id": "ZZZ2",
                    "labels": {
                        "bundle-uuid": "aaa",
                        "bundle-version": "yyy",
                        "caas-collection-name": "lira-prod",
                        "cromwell-workflow-id": "cromwell-ZZZ2",
                        "project_shortname": "project name",
                        "project_uuid": "eee",
                        "workflow-name": "AdapterSmartSeq2SingleCell",
                        "workflow-version": "smartseq2_v2.1.0"
                    },
                    "name": "AdapterSmartSeq2SingleCell",
                    "start": "2019-01-06T20:18:18.102Z",
                    "status": "Failed",
                    "submission": "2019-01-06T20:16:30.804Z"
                }
            }
        },
        "bbb": {
            "type": "primary",
            "aws": {
                "dss_presence": true,
                "in_dss_project_search": true,
                "results_bundles": [
                    "bbb1.yyy"
                ]
            },
            "gcp": {
                "dss_presence": true,
                "in_dss_project_search": true,
                "results_bundles": [
                    "bbb1.yyy"
                ]
            },
            "fqid": "bbb.yyy",
            "analysis_workflows": {
                "FFF": {
                    "end": "2019-01-06T20:35:19.533Z",
                    "id": "FFF",
                    "labels": {
                        "bundle-uuid": "bbb",
                        "bundle-version": "yyy",
                        "caas-collection-name": "lira-prod",
                        "cromwell-workflow-id": "cromwell-FFF",
                        "project_shortname": "project name",
                        "project_uuid": "eee",
                        "workflow-name": "AdapterSmartSeq2SingleCell",
                        "workflow-version": "smartseq2_v2.1.0"
                    },
                    "name": "AdapterSmartSeq2SingleCell",
                    "start": "2019-01-06T20:18:18.102Z",
                    "status": "Succeeded",
                    "submission": "2019-01-06T20:16:30.804Z"
                }
            }
        },
        "ccc": {
            "type": "primary",
            "aws": {
                "dss_presence": true,
                "in_dss_project_search": true,
                "results_bundles": [
                    "ccc1.yyy"
                ]
            },
            "gcp": {
                "dss_presence": true,
                "in_dss_project_search": true,
                "results_bundles": [
                    "ccc1.yyy"
                ]
            },
            "fqid": "ccc.yyy",
            "analysis_workflows": {}
        }
    }
}
```

### Some example commands:
- `dcpdig @analysis wf_uuid='f00eaf60-1622-440f-b5ea-40bc23280fbf' -d staging --credentials="xxx_key_for_staging.json" -v -s bundles,projects`
-  `dcpdig @analysis bundle_uuid='da3fc83e-d108-4536-9ce3-bdd25053c316' -d staging --credentials="xxx_key_for_staging.json" -v`
- `analyze-submission "5c2dfb101603f500078b28de" --deployment="prod" --credentials="xxx_key_for_prod.json"` (this requires Ingest promote their HTTPS endpoint to prod)
- `analyze-submission "5c2dfb101603f500078b28de" --deployment="prod" --credentials="xxx_key_for_prod.json" --fresh` (this requires Ingest promote their HTTPS endpoint to prod)

**Note:** if you want to actually play with the above commands, please let me know, I could guide you obtain the creadentials.

### Limitations and to-dos:
- The analysis agent is consuming a Google Cloud Service Account JSON Key file that has read access to the Secondary Analysis, which is not easy to distribute to developers and users. An OAuth method being invoked from the command line is the ideal way to go. We need to implement this ASAP, otherwise not only the analysis agent is tricky to use, but the Ingest agent will expire in a few days.
- The analysis agent is using block synchronous requests talking to analysis service, this fine with `dcpdig` command but will be extremely slow in the `analyze-submission` command if you have, let's say more than 2k bundles/workflows to query(will take ~10 mins). We need to use multi-threading methods for analysis in `analyze-submission` command.
- There are some scalability [issues and concerns](https://github.com/broadinstitute/cromwell/issues/3115) with the current analysis workflow execution engine, which forces me to use synchronous calls in `analyze-submission`, which in turn causes the inefficiency problem above. We might need to follow up with the Cromwell team in the future.
- The usage of `colored` is a bit chaotic, we need to standardize the colors (what should be used to show entities, how to highlight duplications or errors) across this tool.
- There are no test cases at the moment. I tried Pytest, which will be an ideal testing framework for this tool, but that's too much work in a single PR, will add test cases later. We need to encourage all DCP component engineers to work on this!
- A more powerful linting tool beyond flake8 will be helpful for collaborators who are new to the development process of this tool.
- Integration to readthedocs.io is a must-have to me. But before that, we need to standardize the docstrings used in this code base. I personally prefer Google Style than reST.
- Lack of examples and instructions in the readme. A lot of commands and flags are pretty invisible to those who don't develop this tool. A few Jupyter Notebooks might be a good idea.

### Visions:
Once this tool become stable and efficient, we could  
- use this in the dcp-testing suite so that different errors can be investigated more efficiently, states can be tracked in a better way.
- encourage other people with DCP to use this tool to help their daily work interacting with DCP, such as wranglers.
- as a potential backend of an interface that helps visualize the data-lifecycle across DCP. (or at least directly render the persistent JSON file generated by this tool with JavaScript, as a first step)

sampierson committed:
Build the foundation of the secondary analysis components.

rexwangcc committed:
Update the requirements.

rexwangcc committed:
1st iteration of the analysis agents and finders.

rexwangcc committed:
Refacotr existing component entities and implement the analysis workflow entity.

rexwangcc committed:
Implement analysis agent.

rexwangcc committed:
Make sure the existing component finders use absolute import; implement analysis finder.

rexwangcc committed:
Implement analysis logic for dcpdig command

rexwangcc committed:
Implement analysis logic for analyze-submission command.

rhiananthony commented on dcp_diag/component_agents/analysis_agent.py:
```diff
@@ -0,0 +1,171 @@
+        Args:
+            deployment (str): String representing the deployment environment to be queried on. It could be:
+                "dev", "integration", "staging" or "prod".
+            service_account_key (str): Opotional, path to the service account JSON key file, which is required by
```
spelling - Optional, not Opotional

rhiananthony commented on dcp_diag/component_agents/analysis_agent.py:
```diff
@@ -0,0 +1,171 @@
+        Note, due to the open issue: https://github.com/broadinstitute/cromwell/issues/3115, if the result of workflows
+        are more than ~1000, this function will very likely raise an error. The `with_labels` is a flag controlling the
+        behavior of whether to query the workflows asking for the labels in the response, by default it's set to True,
+        so please set it to False if you don't want to risk getting errored responses.
```
erred not errored

rhiananthony commented on scripts/analyze-submission:
```diff
@@ -242,6 +283,110 @@ class AnalyzeSubmission:
+                   f"\tWorkflows are failed     : {self.failed_analysis_workflow_count}/{self.analysis_workflow_count}\n")
+            if self.errors:
+                # list the errors due to connection so user won't be confusing
+                output(f"\r\tErrored fetch requests   : {self.errors_count}/{self.analysis_workflow_count}\n")
```
erred not errored; or 'Fetch Requests with errors'

rhiananthony commented on scripts/analyze-submission:
```diff
@@ -66,10 +70,44 @@ class AnalyzeSubmission:
+
+        @property
+        def succeeded_analysis_workflow_count(self):
+            """Return the number of succeeded analysis workflows from bundle_map, if the map is empty, return 0."""
```
PEP-8: Line too long

rhiananthony commented on scripts/analyze-submission:
```diff
@@ -66,10 +70,44 @@ class AnalyzeSubmission:
+
+        @property
+        def failed_analysis_workflow_count(self):
+            """Return the number of failed or aborted analysis workflows from bundle_map,
```
PEP-8: Line too long

rhiananthony commented on scripts/analyze-submission:
```diff
@@ -66,10 +70,44 @@ class AnalyzeSubmission:
+
+        @property
+        def running_analysis_workflow_count(self):
+            """Return the number of ongoing analysis workflows from bundle_map, if the map is empty, return 0."""
```
PEP-8: Line too long

rhiananthony commented on scripts/analyze-submission:
```diff
@@ -66,10 +70,44 @@ class AnalyzeSubmission:
+        @property
+        def running_analysis_workflow_count(self):
+            """Return the number of ongoing analysis workflows from bundle_map, if the map is empty, return 0."""
+            return sum([wf_body['status'] not in ('Failed', 'Aborted', 'Succeeded') for (k, v) in self.bundle_map.items() if v['type'] == 'primary' for wf_id, wf_body in v['analysis_workflows'].items()])
```
PEP-8: Line too long

rhiananthony commented on scripts/analyze-submission:
```diff
@@ -242,6 +283,110 @@ class AnalyzeSubmission:
+                all_analysis_workflows_summary_ids = self._get_workflows_summary(analysis_agent=analysis)
+
+                # figure out skippable workflows and exclude them from query to save I/O
+                self.remaining_analysis_workflows_summary = all_analysis_workflows_summary_ids - self.succeeded_workflows
```
PEP-8: Line too long

rhiananthony commented on scripts/analyze-submission:
```diff
@@ -242,6 +283,110 @@ class AnalyzeSubmission:
+            #       workflows ran for the same bundle. Similar to the PHASE 5
+
+            # TODO: make it print the workflow ids when user wants verbosity (V_GOOD_DETAIL)
+            output(f"\r\tWorkflows are succeeded  : {self.succeeded_analysis_workflow_count}/{self.analysis_workflow_count}\n"
```
PEP-8: Line too long

rhiananthony commented on scripts/analyze-submission:
```diff
@@ -357,10 +506,21 @@ class AnalyzeSubmission:
+            output("\nPHASE 4: No auth information provided, skip checking Secondary Analysis for workflows.\n")
+
+        output("\nPHASE 5: Check DSS for secondary bundles:\n", V_SUMMARY)
+        checker5 =self.SearchDSSforSecondaryBundles(deployment=self.deployment, state=self.state, options=args)
```
PEP-8: whitespace after the = sign

rhiananthony commented on dcp_diag/component_entities/analysis_entities.py:
```diff
@@ -0,0 +1,139 @@
+
+    def __eq__(self, other):
+        return (isinstance(self, type(other)) and isinstance(other, type(self)) and (
+        self.uuid, self.name, self.status, self.start_time, self.end_time) == (
```
PEP-8 continuation line (starting at self.) should be indented

rhiananthony reviewed (changes requested):
A few changes requested, but I could also approve without

samanehsan commented on scripts/analyze-submission:
```diff
@@ -66,10 +70,48 @@ class AnalyzeSubmission:
+        @property
+        def running_analysis_workflow_count(self):
+            """Return the number of ongoing analysis workflows from bundle_map, if the map is empty, return 0."""
+            return sum([wf_body['status'] not in ('Failed', 'Aborted', 'Succeeded')
```
Doing a more explicit check that `wf_body['status'] in ('Running', 'Aborting')` would be clearer here. Also where do you think On Hold workflows should go? Right now they will be reflected in this count of running workflows.

rexwangcc replied:
I was intentionally making this "running_analysis_workflow_count" inconsistent with our "Cromwell" status "Running", for not exposing too many details to the end user of dcp-diag and not make the statuses too granular here. The "Running" here is more like "ongoing, but could later go to either succeeded or failed" workflows(which covers Running, Aborting, Submitted, On Hold). So "On Hold" should be reflected here. I could use another word like "ongoing" here to distinguish, what do you think?

rexwangcc replied:
Done renaming

rexwangcc replied:
I'm about to merging this PR, let's try to improve this in a followup PR if it's still not optimal?

samanehsan replied:
Ok, I see what you mean! My reasoning was that users would need to know the possible Cromwell statuses to know what is included in "Running" workflows but with a broader interpretation that it refers to anything "in progress" the specific status details may not be too important.

rexwangcc replied:
@samanehsan Users will still be able to see the detailed exact statuses in the saved result JSON file, but will only see this broader "in progress" in the command line output, not sure if this makes sense!

samanehsan commented on dcp_diag/component_agents/analysis_agent.py:
```diff
@@ -0,0 +1,171 @@
+        result = response.json()
+        all_workflows = [Workflow(wf) for wf in result['results']]
+        total_count = result['totalResultsCount']
+        return total_count, all_workflows
```
I think it's simpler to just have the function return `all_workflows` and then get the `len(all_workflows)` wherever you need the number of workflows. Same for `query_by_project_uuid`. What do you think?

rexwangcc replied:
You are right! Will change!

sampierson commented on scripts/dcpdig:
```diff
@@ -36,9 +40,23 @@ class DcpDig:
         try:
-            finder = Finder.factory(finder_name=args.component, deployment=self.deployment)
+            # TODO: switch to OAuth to simplify the following conditional blocks
+            if args.credentials:
```
I think there might be a way to eliminate this if statement, without needing OAuth.  If we convert `args` to a dict immediately, then delete things from it as we extract them, we can pass what remains to Finder.factory() as `**args`, e.g.:
```
        args = vars(parser.parse_args())
        self.verbose = args.pop('verbose')

        self.deployment = self._choose_deployment(args)
        print(f"Using deployment {self.deployment}")

        component = args.pop('component')
        if component.startswith('@'):
            component = component[1:]

        entities_to_show = args.pop('show').split(',')

        try:
            expression = args.pop('expression')
            finder = Finder.factory(finder_name=component, deployment=self.deployment, **args)
            entity = finder.find(expression)
```

I have done this in commit https://github.com/HumanCellAtlas/dcp-diag/commit/d58095058870fb7784c6180bf8651f4263dc67e3 on branch https://github.com/HumanCellAtlas/dcp-diag/commits/spierson/upload-int-pks.  Please feel free to cherry pick this commit.

sampierson reviewed (approved):
Great work!  Some comments:

* These days I always try to divide my PRs into cleanup and functionality.  Git has a weakness that if you move a file, and change it, you can't tell what the changes were as everything looks new.  For this reason I always do moves/cleanup in one PR and then changes in a second.
* Nice docstrings!  I will start adding those to the upload code.
* Yes it's time this repo got tests.  I'll have a think about how to do that.
* Take a look at the comment in scripts/dcpdig about removing the ugly if.  I prototyped it on a separate branch but didn't re-test the analysis code.

sampierson committed:
dcpdig: restructure __init__ arg processing to eliminate ugly "if"

Finders much now all take **args.

rexwangcc committed:
Change the workflow status term from running to ongoing for better clarity.

rexwangcc committed:
Refactor the analysis agent to make the methods more readable and consistent.

rexwangcc committed:
Update the analysis finder to eliminate ugly conditional statements.

rexwangcc commented:
I have addressed all of the comments and prepare to merge this PR into master, will make a force push to the feature branch to make the history looks nicer.
